File: app/buildz/tmp/kapt3/stubs/debug/com/bits/bee/bpmc/presentation/ui/buka_kasir/BukaTutupKasirSharedViewModel.java
package com.bits.bee.bpmc.presentation.ui.buka_kasir;

import java.lang.System;

/**
 * Created by aldi on 20/06/22.
 */
@dagger.hilt.android.lifecycle.HiltViewModel()
@kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000R\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\b\n\u0002\b\b\b\u0007\u0018\u00002\u000e\u0012\u0004\u0012\u00020\u0002\u0012\u0004\u0012\u00020\u00030\u0001:\u0001\u001fB?\b\u0007\u0012\u0006\u0010\u0004\u001a\u00020\u0005\u0012\u0006\u0010\u0006\u001a\u00020\u0007\u0012\u0006\u0010\b\u001a\u00020\t\u0012\u0006\u0010\n\u001a\u00020\u000b\u0012\u0006\u0010\f\u001a\u00020\r\u0012\u0006\u0010\u000e\u001a\u00020\u000f\u0012\u0006\u0010\u0010\u001a\u00020\u0011\u00a2\u0006\u0002\u0010\u0012J\u0016\u0010\u0013\u001a\u00020\u00142\u0006\u0010\u0015\u001a\u00020\u00162\u0006\u0010\u0017\u001a\u00020\u0018J\u0006\u0010\u0019\u001a\u00020\u0014J\u0006\u0010\u001a\u001a\u00020\u0014J\u0006\u0010\u001b\u001a\u00020\u0014J\u0006\u0010\u001c\u001a\u00020\u0014J\u0006\u0010\u001d\u001a\u00020\u0014J\u0006\u0010\u001e\u001a\u00020\u0014R\u000e\u0010\n\u001a\u00020\u000bX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\b\u001a\u00020\tX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0006\u001a\u00020\u0007X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0004\u001a\u00020\u0005X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u000e\u001a\u00020\u000fX\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0010\u001a\u00020\u0011X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\f\u001a\u00020\rX\u0082\u0004\u00a2\u0006\u0002\n\u0000\u00a8\u0006 "}, d2 = {"Lcom/bits/bee/bpmc/presentation/ui/buka_kasir/BukaTutupKasirSharedViewModel;", "Lcom/bits/bee/bpmc/presentation/base/BaseViewModel;", "Lcom/bits/bee/bpmc/presentation/ui/buka_kasir/BukaTutupKasirState;", "Lcom/bits/bee/bpmc/presentation/ui/buka_kasir/BukaTutupKasirSharedViewModel$UIEvent;", "getActivePossesUseCase", "Lcom/bits/bee/bpmc/domain/usecase/common/GetActivePossesUseCase;", "getActiveCashierUseCase", "Lcom/bits/bee/bpmc/domain/usecase/common/GetActiveCashierUseCase;", "getActiveBranchUseCase", "Lcom/bits/bee/bpmc/domain/usecase/common/GetActiveBranchUseCase;", "bukaKasirUseCase", "Lcom/bits/bee/bpmc/domain/usecase/buka_kasir/BukaKasirUseCase;", "tutupKasirUseCase", "Lcom/bits/bee/bpmc/domain/usecase/tutup_kasir/TutupKasirUseCase;", "getCounterShiftUseCase", "Lcom/bits/bee/bpmc/domain/usecase/buka_kasir/GetCounterShiftUseCase;", "getUserByIdUseCase", "Lcom/bits/bee/bpmc/domain/usecase/rekap_sesi/GetUserByIdUseCase;", "(Lcom/bits/bee/bpmc/domain/usecase/common/GetActivePossesUseCase;Lcom/bits/bee/bpmc/domain/usecase/common/GetActiveCashierUseCase;Lcom/bits/bee/bpmc/domain/usecase/common/GetActiveBranchUseCase;Lcom/bits/bee/bpmc/domain/usecase/buka_kasir/BukaKasirUseCase;Lcom/bits/bee/bpmc/domain/usecase/tutup_kasir/TutupKasirUseCase;Lcom/bits/bee/bpmc/domain/usecase/buka_kasir/GetCounterShiftUseCase;Lcom/bits/bee/bpmc/domain/usecase/rekap_sesi/GetUserByIdUseCase;)V", "doBukaKasir", "Lkotlinx/coroutines/Job;", "modal", "Ljava/math/BigDecimal;", "sesi", "", "doTutupKasir", "getActiveBranch", "getActiveCashier", "getActivePosses", "getUser", "setShift", "UIEvent", "app_debug"})
public final class BukaTutupKasirSharedViewModel extends com.bits.bee.bpmc.presentation.base.BaseViewModel<com.bits.bee.bpmc.presentation.ui.buka_kasir.BukaTutupKasirState, com.bits.bee.bpmc.presentation.ui.buka_kasir.BukaTutupKasirSharedViewModel.UIEvent> {
    private final com.bits.bee.bpmc.domain.usecase.common.GetActivePossesUseCase getActivePossesUseCase = null;
    private final com.bits.bee.bpmc.domain.usecase.common.GetActiveCashierUseCase getActiveCashierUseCase = null;
    private final com.bits.bee.bpmc.domain.usecase.common.GetActiveBranchUseCase getActiveBranchUseCase = null;
    private final com.bits.bee.bpmc.domain.usecase.buka_kasir.BukaKasirUseCase bukaKasirUseCase = null;
    private final com.bits.bee.bpmc.domain.usecase.tutup_kasir.TutupKasirUseCase tutupKasirUseCase = null;
    private final com.bits.bee.bpmc.domain.usecase.buka_kasir.GetCounterShiftUseCase getCounterShiftUseCase = null;
    private final com.bits.bee.bpmc.domain.usecase.rekap_sesi.GetUserByIdUseCase getUserByIdUseCase = null;
    
    @javax.inject.Inject()
    public BukaTutupKasirSharedViewModel(@org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.usecase.common.GetActivePossesUseCase getActivePossesUseCase, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.usecase.common.GetActiveCashierUseCase getActiveCashierUseCase, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.usecase.common.GetActiveBranchUseCase getActiveBranchUseCase, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.usecase.buka_kasir.BukaKasirUseCase bukaKasirUseCase, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.usecase.tutup_kasir.TutupKasirUseCase tutupKasirUseCase, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.usecase.buka_kasir.GetCounterShiftUseCase getCounterShiftUseCase, @org.jetbrains.annotations.NotNull()
    com.bits.bee.bpmc.domain.usecase.rekap_sesi.GetUserByIdUseCase getUserByIdUseCase) {
        super();
    }
    
    @org.jetbrains.annotations.NotNull()
    public final kotlinx.coroutines.Job getActivePosses() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    public final kotlinx.coroutines.Job getUser() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    public final kotlinx.coroutines.Job getActiveBranch() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    public final kotlinx.coroutines.Job getActiveCashier() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    public final kotlinx.coroutines.Job setShift() {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    public final kotlinx.coroutines.Job doBukaKasir(@org.jetbrains.annotations.NotNull()
    java.math.BigDecimal modal, int sesi) {
        return null;
    }
    
    @org.jetbrains.annotations.NotNull()
    public final kotlinx.coroutines.Job doTutupKasir() {
        return null;
    }
    
    @kotlin.Metadata(mv = {1, 7, 1}, k = 1, d1 = {"\u0000\f\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0002\b6\u0018\u00002\u00020\u0001B\u0007\b\u0004\u00a2\u0006\u0002\u0010\u0002\u00a8\u0006\u0003"}, d2 = {"Lcom/bits/bee/bpmc/presentation/ui/buka_kasir/BukaTutupKasirSharedViewModel$UIEvent;", "", "()V", "app_debug"})
    public static abstract class UIEvent {
        
        private UIEvent() {
            super();
        }
    }
}

File: app/src/main/java/com/bits/bee/bpmc/presentation/ui/buka_kasir/BukaTutupKasirSharedViewModel.kt
package com.bits.bee.bpmc.presentation.ui.buka_kasir

import android.content.Context
import androidx.lifecycle.viewModelScope
import com.bits.bee.bpmc.R
import com.bits.bee.bpmc.domain.usecase.buka_kasir.BukaKasirUseCase
import com.bits.bee.bpmc.domain.usecase.buka_kasir.GetCounterShiftUseCase
import com.bits.bee.bpmc.domain.usecase.common.*
import com.bits.bee.bpmc.domain.usecase.rekap_sesi.GetUserByIdUseCase
import com.bits.bee.bpmc.domain.usecase.tutup_kasir.TutupKasirUseCase
import com.bits.bee.bpmc.presentation.base.BaseViewModel
import com.bits.bee.bpmc.utils.BPMConstants
import com.bits.bee.bpmc.utils.BeePreferenceManager
import com.bits.bee.bpmc.utils.DateFormatUtils
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch
import java.math.BigDecimal
import java.util.*
import javax.inject.Inject

/**
 * Created by aldi on 20/06/22.
 */
@HiltViewModel
class BukaTutupKasirSharedViewModel @Inject constructor(
    private val getActivePossesUseCase: GetActivePossesUseCase,
    private val getLastPossesUseCase: GetLastPossesUseCase,
    private val getActiveCashierUseCase: GetActiveCashierUseCase,
    private val getActiveBranchUseCase: GetActiveBranchUseCase,
    private val getDefaultCrcUseCase: GetDefaultCrcUseCase,
    private val bukaKasirUseCase: BukaKasirUseCase,
    private val tutupKasirUseCase: TutupKasirUseCase,
    private val getCounterShiftUseCase: GetCounterShiftUseCase,
    private val getUserByIdUseCase: GetUserByIdUseCase,
    private val beePreferenceManager: BeePreferenceManager,
    private val getRegUseCase: GetRegUseCase
) : BaseViewModel<BukaTutupKasirState, BukaTutupKasirSharedViewModel.UIEvent>() {

    val posPreferences = beePreferenceManager.posPreferences

    val possesActualCashReg = getRegUseCase(BPMConstants.REG_POSSES_ACTUAL_ENDCASH)

    init {
        state = BukaTutupKasirState()
        getActivePosses()
        getActiveBranch()
        getActiveCashier()
        getDefaultCrc()
        setShift()
    }

    fun getLastPosses(context : Context) = viewModelScope.launch {
        val lastPosses = getLastPossesUseCase().first()
        if(lastPosses != null) {
            val lastDate = DateFormatUtils.formatDateToString(BPMConstants.NEW_DATE_FORMAT, lastPosses.endTime!!)
            val dateNow = DateFormatUtils.formatDateToString(BPMConstants.NEW_DATE_FORMAT, Date())
            if(lastDate != dateNow){
                BeePreferenceManager.saveToPreferences(context, context.getString(R.string.pref_counter_sesi), 1)
            }
            BeePreferenceManager.getDataFromPreferences(context, context.getString(R.string.pref_counter_sesi), 1) as Int
        }
    }

    fun getActivePosses() = viewModelScope.launch {
        getActivePossesUseCase().collect {
            updateState(
                state.copy(
                    activePosses = it
                )
            )
        }
    }

    fun getDefaultCrc() = viewModelScope.launch {
        getDefaultCrcUseCase().collect {
            updateState(
                state.copy(
                    crc = it
                )
            )
        }
    }

    fun getUser() = viewModelScope.launch {
        getUserByIdUseCase.invoke(state.activePosses!!.userId).collect {
            updateState(
                state.copy(
                    user = it
                )
            )
        }
    }

    fun getActiveBranch() = viewModelScope.launch {
        getActiveBranchUseCase().collect {
            updateState(
                state.copy(
                    activeBranch = it
                )
            )
        }
    }

    fun getActiveCashier() = viewModelScope.launch {
        getActiveCashierUseCase().collect {
            updateState(
                state.copy(
                    activeCashier = it
                )
            )
        }
    }

    fun setShift() = viewModelScope.launch {
        getCounterShiftUseCase.invoke().collect {

            updateState(
                state.copy(
                    listCasha = it
                )
            )

        }

//        if (state.listCasha!!.size > 0){
//            val trxDate = DateFormatUtils.formatStringToDate(BPMConstants.DEFAULT_DATE_FORMAT, state.listCasha!!.get(0).trxDate)
//            if (SimpleDateFormat("MMdd").format(Date())compareTo(SimpleDateFormat("MMdd").format(trxDate)) > 0){
//                updateState(
//                    state.copy(
//                        shift = 1
//                    )
//                )
//            }else{
//
//            }
//        }else{
//            updateState(
//                state.copy(
//                    shift = 1
//                )
//            )
//        }
    }

    suspend fun doBukaKasir(modal : String, context: Context) = viewModelScope.launch {
        var shift = 1
        val lastPosses = getLastPossesUseCase().first()
        if(lastPosses != null) {
            val lastDate = DateFormatUtils.formatDateToString(BPMConstants.NEW_DATE_FORMAT, lastPosses.endTime!!)
            val dateNow = DateFormatUtils.formatDateToString(BPMConstants.NEW_DATE_FORMAT, Date())
            if(lastDate != dateNow){
                BeePreferenceManager.saveToPreferences(context, context.getString(R.string.pref_counter_sesi), 1)
            }
            shift = BeePreferenceManager.getDataFromPreferences(context, context.getString(R.string.pref_counter_sesi), 1) as Int
        }
        if (modal.isEmpty() || modal.contains(Regex("[A-Za-z]")))
            msgChannel.send("Modal yang anda masukkan salah!")
        else {
            val modalC = BigDecimal(modal)
            bukaKasirUseCase(modalC, shift, state.activeBranch!!, state.activeCashier!!)
            getActivePosses()
            eventChannel.send(UIEvent.NavigateToPos)
        }
    }

    suspend fun doTutupKasir(setorButa : BigDecimal = BigDecimal.ZERO) {
        state.activePosses?.let {
            if(possesActualCashReg.first()?.value == "1"){
                it.totalActualCash = setorButa
            }
            tutupKasirUseCase(
                posses = it
            )
        }
    }

    fun resetActivePosses() = viewModelScope.launch {
        updateState(
            state.copy(
                activePosses = null
            )
        )
    }

    sealed class UIEvent {
        object NavigateToPos : UIEvent()
    }
}